### PR TITLE
Per-layer KV heads, attention logit capping, MoE per-expert scales, NPU multi-input

### PR DESCRIPTION
**Pull request by ncylich**
## Summary

Model-agnostic infrastructure improvements across kernel, graph, engine, and NPU layers.

### Kernel (`kernel.h`, `kernel_attention.cpp`)
- **Attention logit soft-capping**: New `float logit_cap` parameter on `cactus_attention_f16`. When > 0, clamps scores via `logit_cap * tanh(score / logit_cap)` before softmax. Fast path only dispatches when `logit_cap == 0.0f`

### Graph (`graph.h`, `graph_builder.cpp`, `graph_ops_nn.cpp`)
- **MoE per-expert scales**: `moe_layer()` now accepts an optional `per_expert_scale` node providing per-expert FP16 scale factors applied to route weights
- **MoE configurable activation**: `moe_layer()` now takes `Activation activation` parameter (default SILU) instead of hardcoding
- **Logit cap passthrough**: `attention_masked` → kernel call now forwards `logit_cap` from `OpParams`

### Engine (`engine.h`, `engine_cache.cpp`, `engine_model.cpp`)
- **Per-layer KV heads**: KVCache refactored from single global `num_kv_heads` to per-layer `kv_heads` stored in `LayerCache`. Supports models with heterogeneous attention (e.g., alternating local/global layers with different KV head counts)
- **New Config fields**: `expert_intermediate_size`, `global_head_dim`, `num_global_kv_heads`, `attention_k_eq_v`, `enable_moe_block` — parsed from model config JSON
- **Virtual `get_kv_layer_heads()`**: Models can override to provide per-layer KV head counts

### Audio (`engine_signals.cpp`, `cactus_utils.h`)
- **Hann shifted window**: New `hann_shifted` flag on `SpectrogramConfig` for shifted Hann windows (`i + 0.5` indexing). Applied in HTK spectrogram config

### NPU (`npu_ane.mm`)
- **Cached multi-input CoreML prediction**: Pre-allocated input/output buffers via `preallocateMultiInputBuffersWithOutputName` to avoid per-call MLMultiArray allocation overhead
- **Helper functions**: `copyMLArrayToFP16` / `copyFP16ToMLArray` for type-safe data bridging, used in both single-input and multi-input paths
- **Fix**: Gate cached output array usage on `_multiPredictionOptions` to avoid returning stale data on pre-macOS 14

### Tests (`test_kv_cache.cpp`)
- Updated `KVCache::init` and `update_from_graph` call sites to match new per-layer KV heads API

## Test plan
- [x] Builds clean on macOS (cmake + make)
- [x] `cactus test --exhaustive` passes all 11 golden tests (0 failures)
- [x] All unit tests pass (graph, kernel, KV cache, embedding, STT, VLM, LLM, telemetry)
- [x] NPU/ANE loads correctly with fresh model weights

### CODE REVIEW

File: cactus/npu/npu_ane.mm [hunk 1]
```diff
@@ -281,6 +289,93 @@ - (MLMultiArray*)predictWithInput:(NSString*)inputName
...
+    if (_multiPredictionOptions) {
+        result = [_model predictionFromFeatures:provider options:_multiPredictionOptions error:&error];
+    } else {
+        result = [_model predictionFromFeatures:provider error:&error];
+    }
+    if (!result || error) return 0;
+
+    if (_cachedMultiOutputArray) {
+        return copyMLArrayToFP16(_cachedMultiOutputArray, output);
+    }
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/npu/npu_ane.mm:372]: Fixed in 4434a66 — now gates on `_multiPredictionOptions` so we only use the cached output array when `outputBackings` was actually set.

File: cactus/npu/npu_ane.mm [hunk 2]
```diff
@@ -281,6 +289,93 @@ - (MLMultiArray*)predictWithInput:(NSString*)inputName
...
+    return count;
+}
+
+static void copyFP16ToMLArray(const __fp16* data, size_t count, MLMultiArray* array) {
+    if (array.dataType == MLMultiArrayDataTypeFloat16) {
+        memcpy(array.dataPointer, data, count * sizeof(__fp16));
+    } else {
+        float* dst = (float*)array.dataPointer;
+        for (size_t i = 0; i < count; i++) dst[i] = (float)data[i];
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/npu/npu_ane.mm:309]: The cached arrays are allocated from the model's constraint shapes, so sizes always match. Adding a runtime bounds check for an internal API where shapes are guaranteed by construction would be over-engineering.

File: cactus/engine/engine_cache.cpp [hunk 3]
```diff
@@ -516,10 +519,11 @@ void KVCache::compact_to_windows(const std::vector<size_t>& target_windows) {
...
 
         auto& layer = layer_caches[i];
-        size_t layer_tokens = layer.keys.size() / (num_kv_heads * dim * element_size);
+        size_t layer_kv_heads = get_layer_kv_heads(i);
+        size_t layer_tokens = layer.keys.size() / (layer_kv_heads * dim * element_size);
         if (layer_tokens <= target) continue;
 
-        size_t bytes_per_token = num_kv_heads * dim * element_size;
+        size_t bytes_per_token = layer_kv_heads * dim * element_size;
         size_t keep_bytes = target * bytes_per_token;
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/engine/engine_cache.cpp:527]: The existing `dim == 0` check already skips these layers. A model config with 0 kv_heads and nonzero dim would be invalid — no need to double-guard.

File: cactus/engine/engine_cache.cpp [hunk 4]
```diff
@@ -8,16 +8,16 @@
...
     max_seq_len = max_seq;
-    num_kv_heads = kv_heads;
     precision = model_precision;
     element_size = PrecisionTraits::size_of(precision);
 
     layer_caches.resize(num_layers);
     for (size_t i = 0; i < num_layers; i++) {
         layer_caches[i].head_dim = layer_dims[i];
+        layer_caches[i].kv_heads = layer_kv_heads[i];
     }
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/engine/engine_cache.cpp:21]: Both vectors are constructed internally by `get_kv_layer_dims()` and `get_kv_layer_heads()` which always return vectors of size `config_.num_layers`. Not a user-facing API — sizes are guaranteed by construction.

File: cactus/graph/graph_ops_nn.cpp [hunk 5]
```diff
@@ -361,6 +362,15 @@ void compute_moe_layer_node(GraphNode& node, const std::vector<std::unique_ptr<G
...
     }
 
+    const __fp16* expert_scales_fp16 = nullptr;
+    if (has_per_expert_scale) {
+        const auto& scale_buffer = nodes[node_index_map.at(node.input_ids[base_inputs])]->output_buffer;
+        if (scale_buffer.precision != Precision::FP16) {
+            throw std::runtime_error("moe_layer expects FP16 per_expert_scale");
+        }
+        expert_scales_fp16 = scale_buffer.data_as<__fp16>();
+    }
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/graph/graph_ops_nn.cpp:372]: None of the existing MoE weight tensors (w1, w2, w3) are shape-validated either — adding it just for scales would be inconsistent. The graph builder already validates expert count when constructing the node.

File: cactus/npu/npu_ane.mm [hunk 6]
```diff
@@ -470,78 +553,41 @@ - (MLMultiArray*)predictWithInput:(NSString*)inputName
...
+            }
 
-        NSString* outName = nil;
-        if (!output_name.empty()) {
-            outName = [NSString stringWithUTF8String:output_name.c_str()];
-        } else {
-            NSArray<NSString*>* outputNames = impl.modelDescription.outputDescriptionsByName.allKeys;
-            if (outputNames.count > 0) outName = outputNames[0];
+            copyFP16ToMLArray(input.data, total, array);
+            inputDict[nsName] = [MLFeatureValue featureValueWithMultiArray:array];
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/npu/npu_ane.mm:587]: The cached arrays are allocated from the model's static constraint shapes — they don't change across calls. Shape mismatches can't occur in practice.

File: cactus/graph/graph_ops_nn.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/graph/graph_ops_nn.cpp:372]: Duplicate — addressed above. Consistent with existing MoE weight validation.

File: cactus/engine/engine_cache.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/engine/engine_cache.cpp:21]: Duplicate — addressed above. Internal API with guaranteed sizes.

File: cactus/engine/engine_cache.cpp [hunk 7]
```diff
@@ -492,13 +494,14 @@ void KVCache::remove_token_range(size_t start, size_t count) {
...
-        size_t bytes_per_tok = num_kv_heads * dim * element_size;
+        size_t layer_kv_heads = get_layer_kv_heads(i);
+        size_t bytes_per_tok = layer_kv_heads * dim * element_size;
 
         erase_bytes(layer.keys,   start * bytes_per_tok, count * bytes_per_tok, tail_tokens * bytes_per_tok);
         erase_bytes(layer.values, start * bytes_per_tok, count * bytes_per_tok, tail_tokens * bytes_per_tok);
 
         if (precision == Precision::INT8 && !layer.key_scales.empty()) {
-            size_t scaler_per_tok = num_kv_heads * dim / KV_QUANT_GROUP_SIZE;
+            size_t scaler_per_tok = layer_kv_heads * dim / KV_QUANT_GROUP_SIZE;
```
[the remark this answers is not in the document]
Comment (ncylich) [reply cactus/engine/engine_cache.cpp:504]: Pre-existing inconsistency, not introduced by this PR. In practice dim is always a multiple of KV_QUANT_GROUP_SIZE (32) so both formulas give the same result. Could be unified in a follow-up.

File: cactus/engine/engine.h [hunk 8]
```diff
@@ -663,6 +669,9 @@ class Model {
     virtual std::vector<size_t> get_kv_layer_dims() const {
         return std::vector<size_t>(config_.num_layers, config_.attention_head_dim);
     }
+    virtual std::vector<size_t> get_kv_layer_heads() const {
+        return std::vector<size_t>(config_.num_layers, config_.attention_kv_heads);
```
Comment (FarooqMulla) [inline cactus/engine/engine.h]: get_kv_layer_heads() default implementation returns a uniform vector (config_.num_layers, config_.attention_kv_heads). Models with genuinely heterogeneous layers must override this virtual — but there's no documentation, no assertion, and no test verifying a model with mixed head counts actually overrides it. Without a test, the feature silently falls back to uniform heads.

File: tests/test_kv_cache.cpp [hunk 9]
```diff
@@ -27,7 +27,7 @@ bool test_sliding_window_basic() {
     const size_t sink_size = 4;
 
     KVCache cache;
-    cache.init(num_layers, max_seq, num_kv_heads, std::vector<size_t>(num_layers, head_dim), Precision::INT8);
+    cache.init(num_layers, max_seq, std::vector<size_t>(num_layers, head_dim), std::vector<size_t>(num_layers, num_kv_heads), Precision::INT8);
```
Comment (FarooqMulla) [inline tests/test_kv_cache.cpp]: std::vector<size_t>(num_layers, head_dim) and std::vector<size_t>(num_layers, num_kv_heads) — every test uses the same value repeated across all layers. The per-layer feature has zero test coverage for the non-uniform case. Add a test with different head counts per layer.
